File: MaWPhotos/src/main/java/us/mikeandwan/photos/MawApplication.java
package us.mikeandwan.photos;

import android.app.Application;
import androidx.appcompat.app.AppCompatDelegate;

import timber.log.Timber;
import us.mikeandwan.photos.di.ApplicationComponent;
import us.mikeandwan.photos.di.ApplicationModule;
import us.mikeandwan.photos.di.DaggerApplicationComponent;
import us.mikeandwan.photos.di.DataStorageModule;
import us.mikeandwan.photos.di.PhotoApiModule;
import us.mikeandwan.photos.di.PreferenceModule;
import us.mikeandwan.photos.services.CrashReportingTree;


public class MawApplication extends Application {
    public static final String NOTIFICATION_CHANNEL_ID_NEW_CATEGORIES = "notify_new_categories";
    public static final String NOTIFICATION_CHANNEL_ID_UPLOAD_FILES = "files_uploaded";
    public static final int JOB_ID_UPDATE_CATEGORY = 2;
    public static final int JOB_ID_UPLOAD_FILES = 3;
    private static MawApplication _app;

    private int _notificationCount = 0;
    private ApplicationComponent _applicationComponent;

    @Override
    public void onCreate() {
        super.onCreate();

        _app = this;

        if (BuildConfig.DEBUG) {
            Timber.plant(new Timber.DebugTree());
        } else {
            Timber.plant(new CrashReportingTree());
        }

        AppCompatDelegate.setCompatVectorFromResourcesEnabled(true);

        _applicationComponent = DaggerApplicationComponent.builder()
                .applicationModule(new ApplicationModule(this))
                .dataStorageModule(new DataStorageModule())
                .photoApiModule(new PhotoApiModule())
                .preferenceModule(new PreferenceModule())
                .build();

        _applicationComponent.dataServices().wipeTempFiles();
    }


    public int getNotificationCount() {
        return _notificationCount;
    }


    public void setNotificationCount(int count) {
        _notificationCount = count;
    }


    public ApplicationComponent getApplicationComponent() {
        return _applicationComponent;
    }


    public static MawApplication getInstance() {
        return _app;
    }
}

File: MaWPhotos/src/main/java/us/mikeandwan/photos/MawApplication.kt
package us.mikeandwan.photos

import android.app.Application
import android.os.StrictMode
import androidx.hilt.work.HiltWorkerFactory
import androidx.work.*
import dagger.hilt.android.HiltAndroidApp
import timber.log.Timber
import us.mikeandwan.photos.workers.UpdateCategoriesWorker
import java.util.concurrent.TimeUnit
import javax.inject.Inject

@HiltAndroidApp
class MawApplication : Application(), Configuration.Provider {
    @Inject lateinit var workerFactory: HiltWorkerFactory

    override fun onCreate() {
        super.onCreate()
        instance = this

        if (BuildConfig.DEBUG) {
            Timber.plant(Timber.DebugTree())
            StrictMode.enableDefaults()
        }

        schedulePeriodicRefresh()
    }

    override val workManagerConfiguration: Configuration
        get() = Configuration.Builder()
            .setWorkerFactory(workerFactory)
            .build()

    private fun schedulePeriodicRefresh() {
        val constraints = Constraints.Builder()
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .build()

        val work = PeriodicWorkRequestBuilder<UpdateCategoriesWorker>(4, TimeUnit.HOURS)
            .setConstraints(constraints)
            .build()

        val workManager = WorkManager.getInstance(this)

        workManager.enqueueUniquePeriodicWork(
            UpdateCategoriesWorker.WORK_NAME,
            ExistingPeriodicWorkPolicy.KEEP,
            work
        )
    }

    companion object {
        lateinit var instance: MawApplication
            private set
    }
}